File: cmd/kutup/internal/api/federation.go
package api

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"net/textproto"
	"time"
)

// IncomingShare mirrors backend/handlers/fedproxy.go:IncomingShare.
//
// EncryptedCollectionKey is a sealed-box: the inviting server's user
// encrypted the collection key under our public key. The CLI unwraps it
// with crypto.OpenAnonymous(privateKey, publicKey) to recover the
// shared-collection symmetric key, then decrypts encryptedName with that.
type IncomingShare struct {
	ID                     string    `json:"id"`
	RemoteServer           string    `json:"remoteServer"`
	EncryptedCollectionKey string    `json:"encryptedCollectionKey"`
	EncryptedName          string    `json:"encryptedName"`
	NameNonce              string    `json:"nameNonce"`
	CanUpload              bool      `json:"canUpload"`
	CanDelete              bool      `json:"canDelete"`
	UploadQuotaBytes       *int64    `json:"uploadQuotaBytes"`
	CreatedAt              time.Time `json:"createdAt"`
}

// ListIncomingShares returns the federated shares this user has accepted.
func (c *Client) ListIncomingShares() ([]IncomingShare, error) {
	resp, err := c.get("/fed-proxy/incoming")
	if err != nil {
		return nil, err
	}
	var out []IncomingShare
	return out, decodeJSON(resp, &out)
}

// AddIncomingShareRequest body for POST /fed-proxy/incoming.
type AddIncomingShareRequest struct {
	InviteURL string `json:"inviteUrl"`
}

// AddIncomingShare accepts a federated invite URL of the form
// `https://server-b.example/invite/{token}`. The local server proxies
// the invite-token resolution to the remote.
func (c *Client) AddIncomingShare(inviteURL string) (*IncomingShare, error) {
	resp, err := c.postJSON("/fed-proxy/incoming", AddIncomingShareRequest{InviteURL: inviteURL})
	if err != nil {
		return nil, err
	}
	var r IncomingShare
	return &r, decodeJSON(resp, &r)
}

// RemoveIncomingShare deletes the local pointer to a federated share.
// Doesn't notify the remote server (federation today is one-way: the
// owner controls revocation).
func (c *Client) RemoveIncomingShare(shareID string) error {
	resp, err := c.deleteReq("/fed-proxy/incoming/" + shareID)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(body))
	}
	return nil
}

// ProxyListFiles lists files inside a federated share. Returns the same
// FileRow shape as the local /collections/:id/files endpoint — keys are
// wrapped with the federated-share collection key (not the local user's
// master key).
func (c *Client) ProxyListFiles(shareID string) ([]File, error) {
	resp, err := c.get("/fed-proxy/" + shareID + "/files")
	if err != nil {
		return nil, err
	}
	var out []File
	return out, decodeJSON(resp, &out)
}

// ProxyDownload fetches the encrypted file bytes for a file inside a
// federated share. Same encryption-at-rest format as /files/:id/download.
func (c *Client) ProxyDownload(shareID, fileID string) ([]byte, error) {
	resp, err := c.get("/fed-proxy/" + shareID + "/files/" + fileID + "/download")
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(body))
	}
	return io.ReadAll(resp.Body)
}

// ProxyUploadResponse mirrors the remote UploadShareFile response. The
// remote returns either {id: ...} or just a MessageResponse depending on
// build; we accept both — ID may be empty.
type ProxyUploadResponse struct {
	ID string `json:"id"`
}

// ProxyUploadFile posts an encrypted file to a federated share via the
// local /api/fed-proxy/:shareId/upload pass-through. The remote
// UploadShareFile handler enforces:
//   - can_upload permission (returns 403 if not)
//   - share-imposed upload_quota_bytes (atomic FOR UPDATE; 413 on overflow)
//   - sharer's storage_used_bytes is incremented post-write
//
// Multipart shape: same as the local UploadFile — encryptedMetadata,
// metadataNonce, encryptedFileKey, fileKeyNonce, file — minus the
// `collectionId` field (the remote infers it from the share token).
//
// The CLI must already have unwrapped the share's collection key (use
// cli/cmd/share_files.go:resolveSharedCollectionKey) and used it to wrap
// the per-file encFileKey. Asymmetric to local upload only in that one
// step; metadata + content encryption are identical.
func (c *Client) ProxyUploadFile(
	shareID string,
	encryptedMetadata, metadataNonce string,
	encryptedFileKey, fileKeyNonce string,
	encryptedContent []byte,
) (*ProxyUploadResponse, error) {
	var buf bytes.Buffer
	w := multipart.NewWriter(&buf)

	_ = w.WriteField("encryptedMetadata", encryptedMetadata)
	_ = w.WriteField("metadataNonce", metadataNonce)
	_ = w.WriteField("encryptedFileKey", encryptedFileKey)
	_ = w.WriteField("fileKeyNonce", fileKeyNonce)

	h := make(textproto.MIMEHeader)
	h.Set("Content-Disposition", `form-data; name="file"; filename="blob"`)
	h.Set("Content-Type", "application/octet-stream")
	part, err := w.CreatePart(h)
	if err != nil {
		return nil, err
	}
	if _, err := part.Write(encryptedContent); err != nil {
		return nil, err
	}
	w.Close()

	req, err := http.NewRequest(http.MethodPost,
		c.base+"/api/fed-proxy/"+shareID+"/upload", &buf)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", w.FormDataContentType())
	resp, err := c.do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(body))
	}
	// Body may be {id:...} or {message:"ok"}. Best-effort decode — if the
	// shape doesn't include id, we leave it empty and let the caller fall
	// back to listing the share to find the new file.
	body, _ := io.ReadAll(resp.Body)
	r := &ProxyUploadResponse{}
	if len(body) > 0 {
		_ = json.Unmarshal(body, r)
	}
	return r, nil
}

// ProxyDeleteFile removes a file from a federated share. Requires
// can_delete on the share.
func (c *Client) ProxyDeleteFile(shareID, fileID string) error {
	url := c.base + "/api/fed-proxy/" + shareID + "/files/" + fileID
	req, err := http.NewRequest(http.MethodDelete, url, nil)
	if err != nil {
		return err
	}
	resp, err := c.do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(body))
	}
	return nil
}

File: crates/kutup-cli/src/api/federation.rs
//! Incoming Drive shares through the local unified-federation adapter.

use anyhow::Result;
use reqwest::blocking::multipart::{Form, Part};
use reqwest::Method;
use serde::{Deserialize, Serialize};

use super::{Client, File};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct IncomingShare {
    pub id: String,
    #[serde(default)]
    pub remote_domain: String,
    #[serde(default)]
    pub remote_collection_id: String,
    #[serde(default)]
    pub named_share_envelope: String,
    #[serde(default)]
    pub name_envelope: String,
    #[serde(default)]
    pub key_epoch: u32,
    #[serde(default)]
    pub name_revision: u64,
    #[serde(default)]
    pub epoch_statement: String,
    #[serde(default)]
    pub epoch_statement_hash: String,
    #[serde(default)]
    pub owner_user_id: String,
    #[serde(default)]
    pub owner_account: String,
    #[serde(default)]
    pub owner_incarnation_id: String,
    #[serde(default)]
    pub owner_signing_public_key: String,
    #[serde(default)]
    pub owner_authority_public_key: String,
    #[serde(default)]
    pub can_upload: bool,
    #[serde(default)]
    pub can_delete: bool,
    #[serde(default)]
    pub upload_quota_bytes: Option<i64>,
    #[serde(default)]
    pub created_at: String,
}

/// Mirrors the remote `UploadShareFile` response (`{id}` or empty).
#[derive(Debug, Default, Deserialize)]
pub struct ProxyUploadResponse {
    #[serde(default)]
    pub id: String,
}

impl Client {
    /// Lists accepted federated shares. Mirrors `ListIncomingShares`.
    pub fn list_incoming_shares(&self) -> Result<Vec<IncomingShare>> {
        let resp = self
            .request(Method::GET, "/drive/federation/shares")
            .send()?;
        super::decode_json(resp)
    }

    /// Accepts a parsed canonical-domain/capability invite.
    pub fn add_incoming_share(&self, server: &str, capability: &str) -> Result<IncomingShare> {
        let resp = self
            .request(Method::POST, "/drive/federation/shares")
            .header(reqwest::header::CONTENT_TYPE, "application/json")
            .json(&serde_json::json!({ "server": server, "capability": capability }))
            .send()?;
        super::decode_json(resp)
    }

    /// Forgets a local federated-share pointer. Mirrors `RemoveIncomingShare`.
    pub fn remove_incoming_share(&self, share_id: &str) -> Result<()> {
        let resp = self
            .request(
                Method::DELETE,
                &format!("/drive/federation/shares/{share_id}"),
            )
            .send()?;
        super::check_ok(resp)
    }

    /// Lists files inside a federated share. Mirrors `ProxyListFiles`.
    pub fn proxy_list_files(&self, share_id: &str) -> Result<Vec<File>> {
        let resp = self
            .request(
                Method::GET,
                &format!("/drive/federation/shares/{share_id}/files"),
            )
            .send()?;
        super::decode_json(resp)
    }

    /// Streams the encrypted bytes of a file in a federated share (no total
    /// timeout), for the bounded-memory download path.
    pub fn proxy_download_stream(
        &self,
        share_id: &str,
        file_id: &str,
    ) -> Result<reqwest::blocking::Response> {
        let resp = self
            .upload_request(
                Method::GET,
                &format!("/drive/federation/shares/{share_id}/files/{file_id}/content"),
            )
            .send()?;
        if resp.status().as_u16() >= 400 {
            return Err(super::api_error(resp));
        }
        Ok(resp)
    }

    /// Downloads encrypted bytes of a file in a federated share. Mirrors `ProxyDownload`.
    pub fn proxy_download(&self, share_id: &str, file_id: &str) -> Result<Vec<u8>> {
        let resp = self.proxy_download_stream(share_id, file_id)?;
        Ok(resp.bytes()?.to_vec())
    }

    /// Uploads an encrypted file through the signed Drive adapter.
    pub fn proxy_upload_file(
        &self,
        share_id: &str,
        file_id: &str,
        metadata_envelope: &str,
        file_key_envelope: &str,
        encrypted_content: Vec<u8>,
    ) -> Result<ProxyUploadResponse> {
        let part = Part::bytes(encrypted_content)
            .file_name("blob")
            .mime_str("application/octet-stream")?;
        let form = Form::new()
            .text("fileId", file_id.to_string())
            .text("metadataEnvelope", metadata_envelope.to_string())
            .text("fileKeyEnvelope", file_key_envelope.to_string())
            .part("file", part);
        let resp = self
            .request(
                Method::POST,
                &format!("/drive/federation/shares/{share_id}/files"),
            )
            .multipart(form)
            .send()?;
        if resp.status().as_u16() >= 400 {
            return Err(super::api_error(resp));
        }
        Ok(resp.json().unwrap_or_default())
    }
}
